src/rpc/rpc.cc

```diff
@@ -7,25 +7,258 @@
+  buffer += sizeof(msg->msg_seq);
+  total_size += sizeof(msg->msg_seq);
+  // read client_id
+  msg->client_id = *(reinterpret_cast<int32_t*>(buffer));
+  buffer += sizeof(msg->client_id);
+  total_size += sizeof(msg->client_id);
+  // read server_id
+  msg->server_id = *(reinterpret_cast<int32_t*>(buffer));
+  buffer += sizeof(msg->server_id);
+  total_size += sizeof(msg->server_id);
+  // read data size
+  int64_t data_size = *(reinterpret_cast<int64_t*>(buffer));
+  buffer += sizeof(int64_t);
+  total_size += sizeof(int64_t);
+  // read data
+  msg->data.resize(data_size);
+  memcpy(const_cast<char*>(msg->data.data()), buffer, data_size);
+  buffer += data_size;
+  total_size += data_size;
+  // read has_tensor
+  int8_t has_tensor = *(reinterpret_cast<int8_t*>(buffer));
+  buffer += sizeof(int8_t);
+  total_size += sizeof(int8_t);
+  CHECK_EQ(total_size, size);
+  if (has_tensor) {
+    return true;
+  }
+  return false;
+}
+
 RPCStatus SendRPCMessage(const RPCMessage& msg) {
-  // TODO(???): Implement
+  int64_t rpc_meta_size = 0;
+  char* rpc_meta_buffer = SerializeRPCMeta(msg, &rpc_meta_size);
+  network::Message rpc_meta_msg;
+  rpc_meta_msg.data = rpc_meta_buffer;
+  rpc_meta_msg.size = rpc_meta_size;
+  rpc_meta_msg.deallocator = network::DefaultMessageDeleter;
+  CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+    rpc_meta_msg, msg.server_id), ADD_SUCCESS);
+  if (msg.tensors.size() > 0) {
+    std::string zerocopy_blob;
+    StringStreamWithBuffer zc_write_strm(&zerocopy_blob);
+    static_cast<dmlc::Stream *>(&zc_write_strm)->Write(msg.tensors);
+    // send ndarray meta data
+    char* ndarray_meta_buffer = new char[zerocopy_blob.size()];
+    memcpy(ndarray_meta_buffer, zerocopy_blob.data(), zerocopy_blob.size());
+    network::Message ndarray_meta_msg;
+    ndarray_meta_msg.data = ndarray_meta_buffer;
+    ndarray_meta_msg.size = zerocopy_blob.size();
+    ndarray_meta_msg.deallocator = network::DefaultMessageDeleter;
+    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+      ndarray_meta_msg, msg.server_id), ADD_SUCCESS);
+    // send ndarray count
+    char* ndarray_count = new char[sizeof(int32_t)];
+    *(reinterpret_cast<int32_t*>(ndarray_count)) = msg.tensors.size();
+    network::Message ndarray_count_msg;
+    ndarray_count_msg.data = ndarray_count;
+    ndarray_count_msg.size = sizeof(int32_t);
+    ndarray_count_msg.deallocator = network::DefaultMessageDeleter;
+    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+      ndarray_count_msg, msg.server_id), ADD_SUCCESS);
+    // send real ndarray data
+    for (auto ptr : zc_write_strm.buffer_list()) {
+      network::Message ndarray_data_msg;
+      ndarray_data_msg.data = reinterpret_cast<char*>(ptr.data);
+      ndarray_data_msg.size = ptr.size;
+      NDArray tensor = ptr.tensor;
+      ndarray_data_msg.deallocator = [tensor](network::Message*) {};
+      CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+        ndarray_data_msg, msg.server_id), ADD_SUCCESS);
```

Comment:
so the message doesn't own the memory of the tensors. can this cause memory issues? how do we guarantee that the memory isn't free'd before the data is actually sent out.

Reply:
`ndarray_data_msg.deallocator = [tensor](network::Message*) {};` this code own the handler of tensor.

Reply:
ok. i can see it now.